### PR TITLE
[ENT-341] [WIP] Setup enterprise API gateway & service endpoint for Programs.

### PR DESCRIPTION
**Pull request by UmanShahzad**
**Description:** This PR adds an Enterprise API service endpoint for Programs as well as an API gateway. Additionally, it refactors the Course Catalog API wrapper to allow for easily adding more fine-grained service endpoint wrappers from course discovery.

**JIRA:** [ENT-341](https://openedx.atlassian.net/browse/ENT-341)

**Merge deadline:** Coming Monday.

**Testing instructions:**

TBD

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [x] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] All reviewers approved
- [ ] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

**Author concerns:** 

1. I have questions about the implementation which I ask on the upstream ticket.
1. I haven't refactored some of the direct course catalog API tests to be separated out for each new endpoint wrapper.

## TODO

1. Add/remove enterprise/non-enterprise context in serializer, respectively.
1. Add tests for new wrapper(s) / serializer(s).

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=159"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by UmanShahzad** [comment]
@douglashall @brittneyexline @itsjeyd -- if you guys could take an early look at this to let me know if the direction is OK or not, that'd be excellent :) I have some questions on the upstream ticket, too.

File: .travis.yml [hunk 1]
```diff
@@ -23,12 +23,17 @@ matrix:
     - python: 2.7
       env:
         - TOXENV=jasmine
+      before_install:
+        - wget https://github.com/mozilla/geckodriver/releases/download/v0.11.1/geckodriver-v0.11.1-linux64.tar.gz
+        - mkdir geckodriver
+        - tar -xzf geckodriver-v0.11.1-linux64.tar.gz -C geckodriver
+        - export PATH=$PATH:$PWD/geckodriver
```
Comment (UmanShahzad) [inline .travis.yml]: The new jasmine version requires geckodriver for Firefox (actually, it's Selenium that requires Jasmine to require it, afaik). This is currently how jasmine-py is handling it so I just dropped the code here.

**Comment by douglashall** [comment]
@UmanShahzad One thing that would help here is for you to separate different types of changes into separate commits. Maybe one commit for each of these categories:
* General code cleanup
* Requirements upgrades
* API gateway changes
* General code changes which implement the feature

**Comment by UmanShahzad** [comment]
@douglashall Yeah, sorry about the mess of a diff. There was so much overlap going on as I worked through the refactor that it became hard to put things into nice separate commits.

I'll do a reset and try my best to separate things out, and push separate commits again.

**Comment by UmanShahzad** [comment]
@douglashall Take a look and let me know if that's easier on the eyes.

**Comment by douglashall** [comment]
@UmanShahzad Beautiful! Much appreciated. I will start working through this now.

File: enterprise/api/v1/urls.py [hunk 2]
```diff
@@ -1,22 +1,17 @@
...
 """
+
 from __future__ import absolute_import, unicode_literals
 
 from rest_framework.routers import DefaultRouter
 
 from enterprise.api.v1 import views
 
-# Pylint considers module level variable as "Constant" and expects them to be upper-cased
-# that is why we have disabled 'invalid-name' check for variable definition below.
```
Comment (UmanShahzad) [inline enterprise/api/v1/urls.py]: We disable invalid-name in other places, too, and I assume most people know the reason (roughly). Don't need 2 full lines explaining it.

File: test_utils/fake_course_discovery_api.py [hunk 3]
```diff
@@ -15,7 +15,33 @@
     "status": "active",
```
Comment (UmanShahzad) [inline test_utils/fake_course_discovery_api.py]: Not currently utilizing the changes below, but likely will when I write the tests.

File: tox.ini [hunk 4]
```diff
@@ -80,10 +80,6 @@ commands =
 
 [testenv:jasmine]
 passenv = JASMINE_BROWSER DISPLAY
-whitelist_externals =
-    make
-    rm
-    touch
```
Comment (UmanShahzad) [inline tox.ini]: They were unused.

File: api-compact.yaml [hunk 5]
```diff
@@ -10,6 +10,8 @@
 #   /enterprise/api/v1/catalogs/
 #   /enterprise/api/v1/catalogs/{id}/
 #   /enterprise/api/v1/catalogs/{id}/courses/
+#   /enterprise/api/v1/programs/
+#   /enterprise/api/v1/programs/{uuid}/
```
Comment (douglashall) [inline api-compact.yaml]: The endpoints we want to make available are:
```
/enterprise/api/v1/enterprise_catalogs/
/enterprise/api/v1/enterprise_catalogs/{enterprise_catalog_uuid}/
/enterprise/api/v1/program/{uuid}/
```

File: enterprise/api/v1/views.py [hunk 6]
```diff
@@ -308,23 +327,57 @@ def courses(self, request, enterprise_customer, pk=None):  # pylint: disable=inv
...
+        self.ensure_data_exists(request, courses)
+        serializer = serializers.EnterpriseCoursesReadOnlySerializer(courses)
 
         # Add enterprise related context for the courses.
-        serializer.update_enterprise_courses(enterprise_customer, pk)
+        serializer.update_enterprise_courses(enterprise_customer, catalog_id=pk)
+        return get_paginated_response(serializer.data, request)
+
+
+class EnterpriseProgramsViewSet(EnterpriseWrapperApiViewSet):
```
Comment (douglashall) [inline enterprise/api/v1/views.py]: We don't want to make any/all programs in the discovery service available via the Enterprise API. This is where the `EnterpriseCatalog` model comes into play.

File: api-compact.yaml (hunk 5 above)
Comment (UmanShahzad) [inline api-compact.yaml]: Yep, I did this as part of the old requirements. :+1: Of course, we will do `programs` here as noted on the ticket.

File: enterprise/api/v1/views.py (hunk 6 above)
Comment (UmanShahzad) [inline enterprise/api/v1/views.py]: Right, this'll require a rework based off of the new requirements.

**Comment by UmanShahzad** [comment]
@douglashall Ah! He beat me to it :) Yeah, I see now that we have extremely overlapping changes and now I have so many rebase conflicts due to it that it'd be a huge mess to fix.

I'm going to close this PR and start from scratch -- I'll take the important things in, of course.

**Comment by UmanShahzad** [comment]
The new PR I'm currently working on has incorporated whatever was needed from this one, so closing this.

**Pull request closed without merging**